WinRT.Host/WinRT.Host.cpp

```diff
@@ -0,0 +1,396 @@
+        { LibHostAppRootFindFailure, L"The imprinted application path doesn't exist." },
+        { SdkResolverResolveFailure, L"hostfxr_resolve_sdk2 failed to find matching SDK." },
+        { FrameworkCompatFailure, L"The .runtimeconfig.json contains two incompatible framework references." },
+        { FrameworkCompatRetry, L"A previously processed framework reference was reprocessed." },
+        { BundleExtractionFailure, L"Error extracting single-file apphost bundle." },
+        { BundleExtractionIOError, L"Error reading or writing files during single-file apphost bundle extraction." },
+        { LibHostDuplicateProperty, L"The .runtimeconfig.json contains a runtime property which is also produced by the hosting layer." },
+        { HostApiUnsupportedVersion, L"Feature which requires certain version of the hosting layer binaries was used on a version which doesn't support it." },
+        { HostInvalidState, L"The current state is incompatible with the requested operation." },
+        { HostPropertyNotFound, L"A property requested by hostfxr_get_runtime_property_value doesn't exist." },
+        { CoreHostIncompatibleConfig, L"The component being initialized requires framework which is not available" },
+        { HostApiUnsupportedScenario, L"hostfxr doesn't currently support requesting the given delegate type using the given context." },
+    };
+
+    for (auto&& elem : hostfxr_status_messages)
+    {
+        if (elem.status == result)
+        {
+            throw hresult_error(result, elem.message);
+        }
+    }
+    
+    winrt::throw_hresult(result);
+}
+
+inline void check_hostfxr_hresult(hresult const result)
+{
+    if (result < 0)
+    {
+        throw_hostfxr_hresult(result);
+    }
+}
+
+// Using the nethost library, discover the location of hostfxr and get exports
+void load_hostfxr()
+{
+    static const auto is_hostfxr_loaded = [&]()
+    {
+        return(hostfxr_initialize_for_runtime_config &&
+            hostfxr_get_runtime_delegate &&
+            hostfxr_close &&
+            hostfxr_set_error_writer);
+    };
+
+    if (is_hostfxr_loaded())
+    {
+        return;
+    }
+
+    wchar_t buffer[MAX_PATH];
+    size_t buffer_size = sizeof(buffer) / sizeof(wchar_t);
+    check_hostfxr_hresult(get_hostfxr_path(buffer, &buffer_size, nullptr));
+    auto lib = ::LoadLibraryW(buffer);
+    if (lib == 0)
+    {
+        winrt::throw_last_error();
+    }
+
+    hostfxr_initialize_for_runtime_config = (hostfxr_initialize_for_runtime_config_fn)::GetProcAddress(lib, "hostfxr_initialize_for_runtime_config");
+    hostfxr_get_runtime_delegate = (hostfxr_get_runtime_delegate_fn)::GetProcAddress(lib, "hostfxr_get_runtime_delegate");
+    hostfxr_close = (hostfxr_close_fn)::GetProcAddress(lib, "hostfxr_close");
+    hostfxr_set_error_writer = (hostfxr_set_error_writer_fn)::GetProcAddress(lib, "hostfxr_set_error_writer");
+
+    if (!is_hostfxr_loaded())
+    {
+        winrt::throw_last_error();

```

Comment:
Do we expect throw last error to be overridden if any one of those failed then the other after it succeeded.  Or will it be all fail or all pass.

Reply:
ah, good catch